### ISSUE TITLE
making class weight for custom DB

### ISSUE THREAD

**Issue by suhyung** [alexgkendall/caffe-segnet#37]
I am establishing the process for making class weight. 
First, I had checked default DB(CamVid) and default weights in segnet_train.prototxt
As I counted law data pixels of default DB, below data table was shown.

class,  pixels,  files
0   10682767   366
1   14750079   365
2   623349   366
3   20076880   367
4   2845085   349
5   6166762   319
6   743859   351
7   714595   173
8   3719877   360
9   405385   317
10   184967   201
11   2503995   367

and following your paper, calculated weight. left is default weight, right is my result.
f(class) = frequency(class) / (image_count(class) \* 480*360)
weight(class) = median of f(class)) / f(class)

0      0.2595       0.256527308
1      0.1826       0.185282667
2      4.564        4.396287575
3      0.1417       0.136869322
4      0.9051       0.918473131
5      0.3826       0.387319864
6      9.6446       3.533074291
7      1.8418       1.812685267
8      0.6823       0.724619798
9      6.2478       5.855012981
10      7.3614      8.136508447
11                          1.097409921

almost weight values seems similar between them
but only class #6 has big different. Could you confirm my raw data?

Also, I have another question.
If I want to ignore class#5, how can I do it? your reference ignores #11.
so it seems that you described only 0-10 on prototxt file.
but if middle position class should be ignored how can we describe weights? 
and multi classes can be ignored? If it is possible, can you advice the way?
BR

**Comment by yhlleo** [comment]
Have you got the answers? Now, I am confused by this question, too. If you have solved it, would you like to tell me the answer?

**Comment by ArtemSkrebkov** [comment]
@suhyung @yhlleo @alexgkendall 
I'm also interested how authors to get this weight values . I have got weights the same as @suhyung 
Could someone give me an explanation how to get weights as in segnet_train.prototxt?

**Comment by yhlleo** [comment]
@ArtemSkrebkov 
In the paper: [Predicting Depth, Surface Normals and Semantic Labels with a Common Multi-Scale Convolutional Architecture](https://arxiv.org/abs/1411.4734), there are some expainations about the class weight. The SegNet refers to the method, but the results are not same as my caculation results. My results are same as @suhyung's.

**Comment by suhyung** [comment]
@yhlleo @ArtemSkrebkov 
About 1st question, I could not find any solution. anyway, it's glad to see @yhlleo 's calc result.
About 2nd question, I found 2 solutions. 
1. modified the source code for multi ignore function. 
2. modified the DataBase. if I hope to ignore 5th class, changed all images's pixels 5 to 11

**Comment by csjfwang** [comment]
@suhyung @yhlleo @ArtemSkrebkov @pluskid @sergeyk 
Hi, I have the same confuse about the class-weighting .i want to ask , how i should deal with the class-weighting if i want to funtuning the SegNet for a two classes segmentation.Because i don't know how to choose the median of f(class)) and what the ignore_label is. 

Thx and waitting for your reply!!

Jacky_wlwsea

**Comment by suhyung** [comment]
@wlwsea I hope you have already found the answer. In my case, I have converted all unnecessary class data to zero(ignore class).

**Comment by titoghose** [comment]
@suhyung Can you please explain the individual terms of the formula you have used to calculate class weights as I am trying to implement the model on my own dataset and need custom class weights in order to train the model

**Comment by sajjo79** [comment]
Hi,
I am applying segnet on 5 class data. I am not able to understand the role of "ignore label" and the process of weight calculation. any help?

**Comment by mitalbert** [comment]
@suhyung 
I am interested in multi ignore function. Have you tested your implementation? Is it the same as modifying pixel values to 11 in this case? Did you only modify the following files or more?
- ./src/caffe/proto/caffe.proto
- ./include/caffe/layers/softmax_loss_layer.hpp
- ./src/caffe/layers/softmax_loss_layer.cu

Have you uploaded these modifications? If not, would you please share the modified source code for multi ignore function?
